openedx/features/subscriptions/models.py

```diff
@@ -0,0 +1,102 @@
+    )
+
+    subscription_type = models.CharField(choices=SUBSCRIPTION_TYPE_CHOICES, max_length=30)
+    max_allowed_courses = models.PositiveIntegerField(
+        help_text=_(
+            'Maximum number of courses allowed to enroll in paid tracks using the subscription.'
+        ),
+        null=True,
+        blank=True
+    )
+    course_enrollments = models.ManyToManyField(CourseEnrollment, blank=True)
+
+    def __str__(self):
+        return "Subscription of type {subscription_type} with ID {subscription_id} for User {user_email}".format(
+            subscription_type=self.subscription_type,
+            subscription_id=self.subscription_id,
+            user_email=self.user.username,
+        )
+
+    @property
+    def is_active(self):
+        """
+        Get active status of a subscription.
+
+        Check if a subscription is active even if a user can no longer purchase a new course through it.
+        """
+        course_enrollments_count = self.course_enrollments.count()
+        if self.expiration_date and self.max_allowed_courses:
+            return self.expiration_date >= date.today() and course_enrollments_count <= self.max_allowed_courses
+        elif self.expiration_date and not self.max_allowed_courses:
+            return self.expiration_date >= date.today()
+        elif self.max_allowed_courses and not self.expiration_date :
+            return course_enrollments_count <= self.max_allowed_courses
+
+        return True
+
+    @property
+    def is_valid(self):
+        """
+        Get validity status of a subscription.
+
+        Check if a user can enroll in a new course through this subscription.
+        """
+        course_enrollments_count = self.course_enrollments.count()
+        if self.subscription_type == self.LIFETIME_ACCESS:
+            return True
+        elif self.subscription_type == self.FULL_ACCESS_COURSES:
+            return self.expiration_date >= date.today()
+        elif self.subscription_type == self.FULL_ACCESS_TIME_PERIOD:
+            return course_enrollments_count < self.max_allowed_courses
+        elif self.subscription_type == self.LIMITED_ACCESS:
+            return self.expiration_date >= date.today() and course_enrollments_count < self.max_allowed_courses
+
+    @classmethod
+    def get_valid_subscription(self, user_id=None):
+        """
+        Get valid subscription of a user.
+        """
+        filter_params = dict()
+        if user_id:
+            filter_params['user'] = user_id
+
+        user_subscriptions = self.objects.filter(**filter_params)
+        valid_user_subscriptions = filter(lambda subscription: subscription.is_valid, user_subscriptions)
+        valid_user_subscriptions_ids = [subscription.subscription_id for subscription in valid_user_subscriptions]
+        if len(valid_user_subscriptions_ids) > 1:
+            logger.warning('Valid subscription count exceeds 1.')
+            return user_subscriptions.none()
```

Comment:
Is this method will return empty list (in case of multiple valid subscriptions) then the learner can keep buys new subscriptions. We should return the latest valid subscription in this case.

Reply:
So should I just log a warning in case a fault occurs? shouldn't there be a way of letting the system/user know something went wrong?

Reply:
I think for now logging a warning is fine.